### ISSUE TITLE
Unsafe write-free repository cleanup

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#100115]
Today a repository cleanup will first write a new root blob, incrementing its generation, to avoid all risk of cleaning up data written by a concurrent operation which would corrupt the repository. Such concurrent operations would have to be in a different cluster because the various checks in the cluster state ensure that repository cleanup cannot run concurrently to any other snapshot operations within the same cluster.

However, one common application for repository cleanup is because the repository filled up its disk, leaving dangling blobs which can be deleted to free up space. In that case it's not possible to write a new root blob so the cleanup process can't even start. It'd be awfully useful if users could themselves take responsibility for ensuring there are no other clusters concurrently writing to a repository and bypass the root blob write in this situation.

**Comment by ywangd** [comment]
The cleanup process removes two sets of files: stale root blobs and dangling indices. What I wonder is whether we can record these two sets of files (maybe in the repositoryData) when a new snapshot finishes. The cleanup process would only need to read them back without having to write a new file first. Also, I am not quite sure why we are not triggering a cleanup process after finishing a new snapshot? Probably because we don't want too many background cleanups? Maybe we could throttle it?

I guess one issue with my proposal is that disk full is likely encountered when taking a new snapshot. When it happens, it does not have the chance to even write a new repositoryData file. If the existing repositoryData has an empty list of stale files, we are going back to the same problem. It might be possible fix it by adding more phases into snapshot creation. But it does not feel worth the effort.

I was also wondering whether "write-free" cleanup is sufficient. What if there is nothing to cleanup? But I think that's probably *not* a real concern because again "disk full" is likely encountered during snapshoting and will most definitely result into failure and hence dangling files. It is very unlikely that a successful snapshot would happen to consume all available disk bytes.

**Comment by DaveCTurner** [comment]
> What I wonder is whether we can record these two sets of files when a new snapshot finishes.

The trouble is that we may have created some of these files without completing a snapshot. E.g. when we start snapshotting a new index we create a new index UUID, and therefore a directory under `/indices/` but the snapshot may fail before completion so we cannot record its UUID anywhere.

I suppose we could in principle write out a blob with all the new UUIDs before we start to write anywhere, but I'm not sure that buys us anything when we can just list the prefixes under `/indices/`.

> Also, I am not quite sure why we are not triggering a cleanup process after finishing a new snapshot?

Mainly because finishing a new snapshot doesn't add to the data to be cleaned up. It's only if we _don't_ finish the snapshot that we encounter difficulties, but that situation may be because all the masters were restarted and then they wouldn't even remember that there was an ongoing snapshot. Note that we do a cleanup at the end of every delete, so the repository cleanup API is only really useful in situations where the repository is sufficiently broken (e.g. ran out of disk space) that we cannot even delete a snapshot.

**Comment by ywangd** [comment]
Yeah I agree the snapshot creation cannot be relied upon to create the list of stale files since a failed snapshot is where this problem begins with.

Do major blob storage services have reliable timestamp? I am wondering whether it is possible to use latest repository data's timestamp as a cutoff time for "write-free" cleanup, i.e. we only delete older dangling files.

**Comment by DaveCTurner** [comment]
We don't have a portable mechanism for getting timestamps from blobs, but I think that would also not be sufficient since the data to cleanup in a failed snapshot creation attempt will have been written after the last `RepositoryData` blob.

**Comment by ywangd** [comment]
Right. Using timestamp has the same issue that failure happens before the timestamp can be updated which brings us back to the same place. At this point, I think having an opt-in unsafe write-free cleanup is the way to go.

**Comment by DaveCTurner** [comment]
I believe we have consensus that we want to do this so I'm removing the `team-discuss` label.
